File: examples/python/with_openai.py
"""
doredore + OpenAI 統合サンプル

doredore で取得したコンテキストを使って、
OpenAI の GPT モデルで回答を生成します。
"""

import os
from doredore import PyDoredore as Doredore

# OpenAI のインストールが必要: pip install openai
try:
    from openai import OpenAI
except ImportError:
    print("❌ OpenAI ライブラリがインストールされていません")
    print("   pip install openai を実行してください")
    exit(1)


def setup_knowledge_base(rag: Doredore):
    """ナレッジベースをセットアップ"""
    print("📚 ナレッジベースをセットアップ中...\n")

    # コレクション作成
    try:
        rag.create_collection("faq", "よくある質問")
    except:
        pass  # 既に存在する場合

    # サンプルデータ
    knowledge = [
        {
            "content": "永代供養とは、お墓の管理を寺院に委託する供養形態です。継承者がいない方でも安心して利用できます。",
            "metadata": {"category": "永代供養", "importance": "high"}
        },
        {
            "content": "永代供養の費用は、一般的に10万円〜150万円程度です。個別安置期間の長さにより価格が変動します。",
            "metadata": {"category": "料金", "importance": "high"}
        },
        {
            "content": "納骨堂には、ロッカー式、仏壇式、自動搬送式などのタイプがあります。都市部で人気が高まっています。",
            "metadata": {"category": "納骨堂", "importance": "medium"}
        },
        {
            "content": "樹木葬は、墓石の代わりに樹木を墓標とする自然葬の一種です。環境に優しく、費用も比較的安価です。",
            "metadata": {"category": "樹木葬", "importance": "medium"}
        },
    ]

    for item in knowledge:
        try:
            rag.add_document(
                content=item["content"],
                collection="faq",
                metadata=None  # metadata は後で実装
            )
        except Exception as e:
            print(f"⚠️  ドキュメント追加エラー: {e}")

    print("✅ ナレッジベースのセットアップ完了\n")


def chat_with_rag(rag: Doredore, openai_client: OpenAI, question: str):
    """RAG + OpenAI でチャット"""

    print(f"💭 質問: {question}\n")

    # 1. RAG でコンテキストを取得
    print("🔍 関連情報を検索中...")
    enrich_result = rag.enrich(
        query=question,
        collection="faq",
        top_k=3,
        threshold=0.0
    )

    print(f"✅ {len(enrich_result.sources)} 件の関連情報を取得\n")

    # 検索結果を表示
    print("📊 検索結果:")
    for i, source in enumerate(enrich_result.sources, 1):
        print(f"  [{i}] スコア: {source.score:.3f}")
        print(f"      {source.content}\n")

    # 2. OpenAI で回答生成
    print("🤖 AI が回答を生成中...\n")

    try:
        response = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {
                    "role": "system",
                    "content": f"""あなたは親切な供養コンサルタントです。
以下の情報を参考に、ユーザーの質問に丁寧に回答してください。

【参考情報】
{enrich_result.context}

回答は簡潔で分かりやすく、専門用語は説明を加えてください。"""
                },
                {
                    "role": "user",
                    "content": question
                }
            ],
            temperature=0.7,
            max_tokens=500
        )

        answer = response.choices[0].message.content

        print("=" * 60)
        print("💡 回答:")
        print("=" * 60)
        print(answer)
        print("=" * 60)
        print()

        # ソース情報を表示
        print("📚 参考にした情報:")
        for i, source in enumerate(enrich_result.sources, 1):
            print(f"  [{i}] {source.content[:60]}... (スコア: {source.score:.3f})")

        print()

    except Exception as e:
        print(f"❌ OpenAI API エラー: {e}")


def main():
    # 環境変数チェック
    if "OPENAI_API_KEY" not in os.environ:
        print("❌ OPENAI_API_KEY 環境変数が設定されていません")
        print("   export OPENAI_API_KEY='your-api-key' を実行してください")
        return

    # 初期化
    print("🚀 doredore を初期化中...\n")
    rag = Doredore(
        db_path="./knowledge_with_ai.db",
        model="bge-small-en-v1.5"
    )

    openai_client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])

    # ナレッジベースのセットアップ
    setup_knowledge_base(rag)

    # サンプル質問
    questions = [
        "永代供養の費用はどれくらいですか？",
        "樹木葬と納骨堂の違いを教えてください",
        "継承者がいない場合、どうすればいいですか？"
    ]

    for i, question in enumerate(questions, 1):
        print(f"\n{'='*60}")
        print(f"質問 {i}/{len(questions)}")
        print(f"{'='*60}\n")
        chat_with_rag(rag, openai_client, question)

    print("\n🎉 デモ完了！")


if __name__ == "__main__":
    main()

File: examples/nodejs/with_openai.js
#!/usr/bin/env node
/**
 * doredore + OpenAI Integration Example
 *
 * This example shows how to use doredore with OpenAI's GPT models
 * to build a question-answering system with retrieval-augmented generation.
 *
 * Requirements:
 *   npm install openai
 *   export OPENAI_API_KEY="your-api-key-here"
 */

const { Doredore } = require('doredore');

// Check if OpenAI is available
let OpenAI;
try {
  OpenAI = require('openai').default;
} catch (e) {
  console.error('Error: openai package not found.');
  console.error('Please install it with: npm install openai');
  process.exit(1);
}

// Check for API key
if (!process.env.OPENAI_API_KEY) {
  console.error('Error: OPENAI_API_KEY environment variable not set.');
  console.error('Please set it with: export OPENAI_API_KEY="your-api-key"');
  process.exit(1);
}

console.log('='.repeat(60));
console.log('doredore + OpenAI Integration Example');
console.log('='.repeat(60));
console.log();

// ============================================================================
// 1. Initialize Doredore and OpenAI
// ============================================================================

console.log('1. Initializing...');
const rag = new Doredore('./knowledge.db', 'bge-small-en-v1.5');
const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
console.log('   ✅ Doredore and OpenAI initialized!');
console.log();

// ============================================================================
// 2. Prepare Knowledge Base
// ============================================================================

console.log('2. Preparing knowledge base...');

try {
  rag.createCollection('faq', 'よくある質問');
} catch (e) {
  // Collection may already exist
}

const documents = [
  '永代供養とは、お墓の管理を寺院に委託する供養形態です。',
  '永代供養の費用は、一般的に10万円〜150万円程度です。',
  '納骨堂には、ロッカー式、仏壇式、自動搬送式などのタイプがあります。',
  '樹木葬は、墓石の代わりに樹木を墓標とする自然葬の一種です。',
  '一般墓は家族代々で受け継がれますが、永代供養墓は寺院が永続的に管理します。'
];

for (const doc of documents) {
  try {
    rag.addDocument(doc, 'faq');
  } catch (e) {
    // Document may already exist
  }
}

console.log('   ✅ Knowledge base ready!');
console.log();

// ============================================================================
// 3. RAG Question Answering Function
// ============================================================================

async function ragAnswer(question) {
  console.log(`Question: "${question}"`);
  console.log();

  // Step 1: Retrieve relevant context using RAG
  console.log('  [RAG] Retrieving relevant documents...');
  const enrichResult = rag.enrich(question, 'faq', null, 3, 0.3);

  console.log(`  [RAG] Found ${enrichResult.sources.length} relevant documents`);
  for (const [index, source] of enrichResult.sources.entries()) {
    console.log(`    ${index + 1}. Score: ${source.score.toFixed(3)}`);
  }
  console.log();

  // Step 2: Call OpenAI with retrieved context
  console.log('  [LLM] Generating answer with GPT-4...');

  const systemPrompt = `あなたは永代供養に関する質問に答えるアシスタントです。
以下の参考情報を基に、正確かつ簡潔に回答してください。

参考情報:
${enrichResult.context}`;

  const response = await openai.chat.completions.create({
    model: 'gpt-4o-mini',  // or gpt-4
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: question }
    ],
    temperature: 0.7,
    max_tokens: 500
  });

  const answer = response.choices[0].message.content;

  console.log();
  console.log('Answer:');
  console.log('─'.repeat(60));
  console.log(answer);
  console.log('─'.repeat(60));
  console.log();

  return {
    question,
    answer,
    sources: enrichResult.sources,
    tokensUsed: response.usage.total_tokens
  };
}

// ============================================================================
// 4. Run Examples
// ============================================================================

async function main() {
  console.log('3. Running RAG + OpenAI examples...');
  console.log();
  console.log('='.repeat(60));
  console.log();

  // Example 1
  const result1 = await ragAnswer('永代供養とは何ですか？');
  console.log(`Tokens used: ${result1.tokensUsed}`);
  console.log();
  console.log('='.repeat(60));
  console.log();

  // Example 2
  const result2 = await ragAnswer('永代供養の費用はどれくらいですか？');
  console.log(`Tokens used: ${result2.tokensUsed}`);
  console.log();
  console.log('='.repeat(60));
  console.log();

  // Example 3
  const result3 = await ragAnswer('一般墓との違いを教えてください。');
  console.log(`Tokens used: ${result3.tokensUsed}`);
  console.log();

  console.log('='.repeat(60));
  console.log('✅ All examples completed successfully!');
  console.log('='.repeat(60));
}

main().catch(error => {
  console.error('Error:', error.message);
  process.exit(1);
});
